sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py

```diff
@@ -0,0 +1,952 @@
+     AzureKeyCredential type or a TokenCredential type. Required.
+    :type credential: ~azure.core.credentials.AzureKeyCredential or
+     ~azure.core.credentials.TokenCredential
+    :keyword api_version: The API version to use for this operation. Default value is
+     "2024-05-01-preview". Note that overriding this default value may result in unsupported
+     behavior.
+    :paramtype api_version: str
+    :return: The appropriate client associated with the given endpoint
+    :rtype: ~azure.ai.inference.ChatCompletionsClient or ~azure.ai.inference.EmbeddingsClient
+     or ~azure.ai.inference.ImageEmbeddingsClient
+    :raises ~azure.core.exceptions.HttpResponseError
+    """
+
+    with ChatCompletionsClient(endpoint, credential, **kwargs) as client: # Pick any of the clients, it does not matter.
+        model_info = client.get_model_info() # type: ignore
+
+    _LOGGER.info("model_info=%s", model_info)
+    if not model_info.model_type:
+        raise ValueError(
+            "The AI model information is missing a value for `model type`. Cannot create an appropriate client."
+        )
+
+    # TODO: Remove "completions" and "embedding" once Mistral Large and Cohere fixes their model type
+    if model_info.model_type in (_models.ModelType.CHAT, "completion"):
+        chat_completion_client = ChatCompletionsClient(endpoint, credential, **kwargs)
+        chat_completion_client._model_info = model_info  # pylint: disable=protected-access,attribute-defined-outside-init
+        return chat_completion_client
+
+    if model_info.model_type in (_models.ModelType.EMBEDDINGS, "embedding"):
+        embedding_client = EmbeddingsClient(endpoint, credential, **kwargs)
+        embedding_client._model_info = model_info  # pylint: disable=protected-access,attribute-defined-outside-init
+        return embedding_client
+
+    if model_info.model_type == _models.ModelType.IMAGE_EMBEDDINGS:
+        image_embedding_client = ImageEmbeddingsClient(endpoint, credential, **kwargs)
+        image_embedding_client._model_info = model_info  # pylint: disable=protected-access,attribute-defined-outside-init
+        return image_embedding_client
+
+    raise ValueError(f"No client available to support AI model type `{model_info.model_type}`")
+
+
+class ChatCompletionsClient(ChatCompletionsClientGenerated):
+
+    @overload
+    def complete(
+        self,
+        *,
+        messages: List[_models.ChatRequestMessage],
+        content_type: str = "application/json",
+        model_extras: Optional[Dict[str, Any]] = None,
+        frequency_penalty: Optional[float] = None,
+        presence_penalty: Optional[float] = None,
+        temperature: Optional[float] = None,
+        top_p: Optional[float] = None,
+        max_tokens: Optional[int] = None,
+        response_format: Optional[Union[str, _models.ChatCompletionsResponseFormat]] = None,
+        stop: Optional[List[str]] = None,
+        stream: Optional[bool] = None,
```

Comment:
We should overload on literal `True`/`False` values. That way we can correctly type the return value and avoid the Union.

Reply:
Per IM conversation, we stay with one 'complete' method for both streaming and non-streaming. And fixed mypy and pyright errors by adding overloads with `stream: Literal`. Commit: https://github.com/Azure/azure-sdk-for-python/pull/34937/commits/0d03ad2ed5c266deed840ec69a2c2396ef47f5b4